arangod/Cluster/DBServerAgencySync.cpp

```diff
@@ -52,125 +60,176 @@ void DBServerAgencySync::work() {
     
-    result.success = true;  // unless overwritten by actual result
+  } catch (std::exception const& e) {
+    LOG_TOPIC(ERR, Logger::MAINTENANCE)
+      << "Failed to handle plan change: " << e.what();
+  }
 
-    if (res->IsObject()) {
-      v8::Handle<v8::Object> o = res->ToObject();
+  if (rb.isClosed()) {
 
-      v8::Handle<v8::Array> names = o->GetOwnPropertyNames();
-      uint32_t const n = names->Length();
-      
-      for (uint32_t i = 0; i < n; ++i) {
-        v8::Handle<v8::Value> key = names->Get(i);
-        v8::String::Utf8Value str(key);
+    auto report = rb.slice();
+    if (report.isObject()) {
 
-        v8::Handle<v8::Value> value = o->Get(key);
+      std::vector<std::string> agency = {"phaseTwo", "agency"};
+      if (report.hasKey(agency) && report.get(agency).isObject()) {
+      
+        auto phaseTwo = report.get(agency);
+        LOG_TOPIC(DEBUG, Logger::MAINTENANCE)
+          << "DBServerAgencySync reporting to Current: " << phaseTwo.toJson();
+      
+        // Report to current
+        if (!phaseTwo.isEmptyObject()) {
         
-        if (value->IsNumber()) {
-          if (strcmp(*str, "plan") == 0) {
-            result.planVersion =
-              static_cast<uint64_t>(value->ToInteger()->Value());
-          } else if (strcmp(*str, "current") == 0) {
-            result.currentVersion =
-              static_cast<uint64_t>(value->ToInteger()->Value());
+          std::vector<AgencyOperation> operations;
+          for (auto const& ao : VPackObjectIterator(phaseTwo)) {
+            auto const key = ao.key.copyString();
+            auto const op = ao.value.get("op").copyString();
+            if (op == "set") {
+              auto const value = ao.value.get("payload");
+              operations.push_back(
+                AgencyOperation(key, AgencyValueOperationType::SET, value));
+            } else if (op == "delete") {
+              operations.push_back(
+                AgencyOperation(key, AgencySimpleOperationType::DELETE_OP));
+            }
+          }
+          operations.push_back(
+            AgencyOperation(
+              "Current/Version", AgencySimpleOperationType::INCREMENT_OP));
+          AgencyWriteTransaction currentTransaction(operations);
+          AgencyCommResult r = comm.sendTransactionWithFailover(currentTransaction);
+          if (!r.successful()) {
+            LOG_TOPIC(ERR, Logger::MAINTENANCE) << "Error reporting to agency";
+          } else {
+            LOG_TOPIC(DEBUG, Logger::MAINTENANCE) << "Invalidating current in ClusterInfo";
+            clusterInfo->invalidateCurrent();
           }
-        } else if (value->IsBoolean() && strcmp(*str, "success")) {
-          result.success = TRI_ObjectToBoolean(value);
         }
       }
-    } else {
-      LOG_TOPIC(ERR, Logger::HEARTBEAT)
-        << "handlePlanChange returned a non-object";
-      return result;
+    
+      result = DBServerAgencySyncResult(
+        tmp.ok(),
```

Comment:
If comm.sendTransactionWithFailover() fails, the result is ok() based upon phaseTwo()'s execution?

Reply:
Good point. Added a FIXMEMAINTENANCE